cdc/redo/writer/file.go

```diff
@@ -403,72 +404,113 @@ func (w *Writer) rotate() error {
-	}
+// deleteFilesInStorages delete files in both local storage and s3, delete in s3 first, if success try to delete in local.
+// if delete in local then s3, s3 fail the file will still exist in s3.
+// for normal case should be fine, but in changefeed remove case, if new created changefeed with the same name,
+// the file in s3 will belong to the new one, which may cause error if used to apply for recovery.
+func (w *Writer) deleteFilesInStorages(remove []os.FileInfo) error {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultS3Timeout)
+	defer cancel()
 
-	if w.cfg.S3Storage {
-		// since if fail delete in s3, do not block any path, so just log the error if any
-		go func() {
-			var errs error
-			for _, f := range remove {
-				err := w.storage.DeleteFile(context.Background(), f.Name())
-				errs = multierr.Append(errs, err)
-			}
-			if errs != nil {
-				errs = cerror.WrapError(cerror.ErrS3StorageAPI, errs)
-				log.Warn("delete redo log in s3 fail", zap.Error(errs))
+	eg, eCtx := errgroup.WithContext(ctx)
+	for _, f := range remove {
+		name := f.Name()
+		eg.Go(func() error {
+			err := w.storage.DeleteFile(eCtx, name)
+			if err != nil {
+				// if fail then retry, may end up with notExit err, ignore the error
+				if !isNotExistInS3(err) {
+					return cerror.WrapError(cerror.ErrS3StorageAPI, err)
+				}
 			}
-		}()
-	}
 
-	return nil
+			err = os.Remove(filepath.Join(w.cfg.Dir, name))
+			return cerror.WrapError(cerror.ErrRedoFileOp, err)
+		})
+	}
+	return eg.Wait()
 }
 
 // shouldRemoved remove the file which commitTs in file name (max commitTs of all event ts in the file) < checkPointTs,
 // since all event ts < checkPointTs already sent to sink, the log is not needed any more for recovery
-func (w *Writer) shouldRemoved(checkPointTs uint64, f os.FileInfo) (bool, error) {
-	if filepath.Ext(f.Name()) != common.LogEXT {
+// if checkPointTs == gcAllTs means remove all logs written by the fileWriter
+func (w *Writer) shouldRemoved(checkPointTs *big.Int, name string) (bool, error) {
+	ext := filepath.Ext(name)
+	tsShouldRM := false
+
+	if checkPointTs.IsInt64() {
+		// gcAllTs ==-1 by default
+		if checkPointTs.Int64() < gcAllTs {
+			return false, cerror.ErrRedoCheckPointTsInvalid.GenWithStackByArgs(checkPointTs.String(), gcAllTs)
+		}
+		if checkPointTs.Int64() == gcAllTs {
+			tsShouldRM = true
+			// may exist .tmp file cause by not closed safely
+			if filepath.Ext(name) == common.TmpEXT {
+				name = strings.TrimSuffix(name, common.TmpEXT)
+				ext = filepath.Ext(name)
+			}
+		}
+	}
+
+	if ext != common.LogEXT {
 		return false, nil
 	}
 
-	commitTs, fileType, err := common.ParseLogFileName(f.Name())
+	commitTs, fileType, err := common.ParseLogFileName(name)
 	if err != nil {
 		return false, err
 	}
 
-	return commitTs < checkPointTs && fileType == w.cfg.FileType, nil
+	return (tsShouldRM || commitTs < checkPointTs.Uint64()) && fileType == w.cfg.FileType, nil
```

Comment:
It might be better if we could explain it, or break it down into its corresponding variables to illustrate it!